### PR TITLE
Aggregate prometheus metrics into one file on worker death.

### PR DESCRIPTION
**Pull request by bloodearnest**
This should stop the leaking files and subsequent increasing scrape
response times.

### CODE REVIEW

File: talisker/metrics.py [hunk 1]
```diff
@@ -112,3 +121,126 @@ def inc(self, amount=1, **labels):
...
+
+    all_files = [histogram_archive, counter_archive] + list(files)
+    tmp = tempfile.mkdtemp(prefix='prometheus_aggregate')
+    try:
+        for filename in all_files:
+            path = os.path.join(prom_dir, filename)
+            if os.path.exists(path):
+                shutil.copy(path, tmp)
+
+        collector = multiprocess.MultiProcessCollector(CollectorRegistry())
```
Comment (maxiberta) [inline talisker/metrics.py]: Shouldn't `collector` receive the `tmp` path as a parameter? Or maybe I'm missing something.

File: tests/conftest.py [hunk 2]
```diff
@@ -29,16 +29,22 @@
 
 from builtins import *  # noqa
 
+# make sure prometheus is setup in multiprocess mode. We don't actuall use this
+# dir in tests, as each test gets it's own directory, but this ensures
```
Comment (maxiberta) [inline tests/conftest.py]: s/it's/its/.

File: talisker/endpoints.py [hunk 3]
```diff
@@ -123,6 +123,17 @@ def ok_response():
     return Response(str(talisker.revision.get()) + '\n')
 
 
+@module_cache
+def test_counter():
+    # this is created lazily as we may not have prometheus installed.
+    # Also, isolate in it's own registry. Multiprocess mode doesn't really use
+    # registries, but it helps keep things separated, which is useful in
+    # testing
```
Comment (sparkiegeek) [inline talisker/endpoints.py]: missing full stop

File: talisker/metrics.py [hunk 4]
```diff
@@ -132,3 +139,175 @@ def inc(self, amount=1, **labels):
...
+
+def histogram_sorter(sample):
+    # sort histogram samples in order of bucket size
+    name, labels, _ = sample
+    return name, float(labels.get('le', 0))
+
+
+def prometheus_cleanup_worker(pid):
+    """Clean up after a multiprocess worker has died."""
+    if prometheus_client is None:
```
Comment (sparkiegeek) [inline talisker/metrics.py]: what's this guarding against? If prometheus_client fails to import it's set to `False`, not `None`

File: talisker/metrics.py [hunk 5]
```diff
@@ -132,3 +139,175 @@ def inc(self, amount=1, **labels):
...
+            tmp_histogram.name, os.path.join(prom_dir, histogram_archive))
+        os.rename(
+            tmp_counter.name, os.path.join(prom_dir, counter_archive))
+
+        for path in paths:
+            if os.path.exists(path):
+                os.unlink(path)
+
+    except Exception:
+        # we should never fail at cleaning up
```
Comment (sparkiegeek) [inline talisker/metrics.py]: agree with the sentiment, let's put it in the caller though? i.e. `prometheus_multiprocess_worker_exit`

File: talisker/metrics.py [hunk 6]
```diff
@@ -132,3 +139,175 @@ def inc(self, amount=1, **labels):
...
+        prom_dir = os.environ['prometheus_multiproc_dir']
+        worker_files = [
+            'histogram_{}.db'.format(pid),
+            'counter_{}.db'.format(pid),
+        ]
+        paths = [os.path.join(prom_dir, f) for f in worker_files]
+
+        # check at least one worker file exists
+        if not any(os.path.exists(path) for path in paths):
+            return
```
Comment (sparkiegeek) [inline talisker/metrics.py]: how about putting the `os.path.exists()` check in the list-comp for `paths`?

That way we can just check `if not paths` here, and drop the check on `unlink` too?

File: talisker/metrics.py [hunk 7]
```diff
@@ -132,3 +139,175 @@ def inc(self, amount=1, **labels):
...
+                    buckets[without_le][bucket[0]] += value
+                else:
+                    # _sum/_count
+                    samples[(name, labels)] += value
+
+            else:
+                # Counter and Summary.
+                samples[(name, labels)] += value
+
+        # modified to remove accumulation
```
Comment (sparkiegeek) [inline talisker/metrics.py]: let's explicitly mark the EOVERBATIM?

File: talisker/metrics.py [hunk 8]
```diff
@@ -132,3 +139,175 @@ def inc(self, amount=1, **labels):
...
+
+    1. it takes its files as an argument, rather than hardcoding '*.db', an
+       skips none existent files
+    2. it does not accumulate histograms, as it is intended for writing that
+       data back out
+    3. it uses an OrderedDict to preserve label order, and thus metric identity
+
+    It needs to be kept up to date with changes to prometheus_client as much as
+    possible, or until changes are landed upstream to allow better reuse.
+    """
```
Comment (sparkiegeek) [inline talisker/metrics.py]: WDYT about an explicit whitelist and failure if we're using a version of prometheus_client that's not been checked?

It means we have to explicitly double check this when we bump prometheus_client but I think that's a Good Thing:tm:

File: talisker/metrics.py [hunk 9]
```diff
@@ -132,3 +139,175 @@ def inc(self, amount=1, **labels):
...
+        # we should never fail at cleaning up
+        logger.exception('failed to cleanup prometheus worker files')
+
+
+def collect(files):
+    """This almost verbatim from MultiProcessCollector.collect().
+
+    It differs in a few ways:
+
+    1. it takes its files as an argument, rather than hardcoding '*.db', an
```
Comment (sparkiegeek) [inline talisker/metrics.py]: s/an/and/

File: tests/conftest.py [hunk 10]
```diff
@@ -29,16 +29,22 @@
 
 from builtins import *  # noqa
 
+# make sure prometheus is setup in multiprocess mode. We don't actuall use this
```
Comment (sparkiegeek) [inline tests/conftest.py]: *actually

File: tests/test_metrics.py [hunk 11]
```diff
@@ -29,69 +29,83 @@
...
         statsd='{name}.{label}',
+        registry=registry,
     )
 
     labels = {'label': 'value'}
-    count = get_metric('test_histogram_count', **labels)
-    sum = get_metric('test_histogram_sum', **labels)
-    bucket = get_metric('test_histogram_bucket', le='2.5', **labels)
+    count = registry.get_metric('test_histogram_count', **labels)
+    sum = registry.get_metric('test_histogram_sum', **labels)
```
Comment (sparkiegeek) [inline tests/test_metrics.py]: let's not shadow the builtin?

File: talisker/metrics.py (hunk 4 above)
Comment (bloodearnest) [reply talisker/metrics.py]: It's just defence in depth, but I've fixed the check

File: talisker/metrics.py (hunk 6 above)
Comment (bloodearnest) [reply talisker/metrics.py]: Well, then I have to join twice.
If could do it in two steps, though, will try that

File: talisker/metrics.py [hunk 12]
```diff
@@ -45,6 +50,8 @@
 except ImportError:
     statsd = False
 
+logger = logging.getLogger(__name__)
```
Comment (maxiberta) [inline talisker/metrics.py]: This looks like a duplicate line.

File: talisker/metrics.py (hunk 8 above)
Comment (bloodearnest) [reply talisker/metrics.py]: I think I would probably prefer to just issue a warning on startup. We can't control what versions of prometheus_client is used in the main project, as I wouldn't want to error or block if they it goes beyond what talisker supports.

Comment (maxiberta) [reply talisker/metrics.py]: This code duplication looks dangerous to me; I'd rather fork `prometheus_client` instead. That said, can't why "unaccumulate" results as in previous versions of this branch?

Comment (bloodearnest) [reply talisker/metrics.py]: Once we've validated this approach in prod, I intend to propose some small changes upstream that will mean this duplication can disappear, time permitting.

Comment (bloodearnest) [reply talisker/metrics.py]: @maxiberta we can unaccumlate, and I did, but we cannot preserve label order without duplicating, and that is essential.

I don't think forking the whole library is a better solution that a custom collection function, that is only used for one specific purpose.

Comment (bloodearnest) [reply talisker/metrics.py]: If you look at the first commit, you can see an approach that did not duplicate. I got it 90% of the way there, only to fail at the final hurdle.

Comment (bloodearnest) [reply talisker/metrics.py]: ultimately, prometheus scraping is not sensitive to label order, but the mmaped file format parsing is

**Comment by bloodearnest** [comment]
This pull request **introduces 2 alerts** and **fixes 1** when merging 3e684dc20a5f77ba982c6303cc361156068d078d into dabd40b8b4a5a2a630661cbad960ca18bdfb5143 - [view on LGTM.com](https://lgtm.com/projects/g/canonical-ols/talisker/rev/pr-eb4a308134d937232dc45207917b9d51f83fb874)

**new alerts:**

* 2 for Module is imported with &#39;import&#39; and &#39;import from&#39;

**fixed alerts:**

* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by bloodearnest** [comment]
This pull request **introduces 2 alerts** and **fixes 1** when merging c14f4e3cef492cadfebd2b94a8539511d798a0eb into dabd40b8b4a5a2a630661cbad960ca18bdfb5143 - [view on LGTM.com](https://lgtm.com/projects/g/canonical-ols/talisker/rev/pr-60c33609e555175a76e0cf345ae2fcc19569a7d6)

**new alerts:**

* 2 for Module is imported with &#39;import&#39; and &#39;import from&#39;

**fixed alerts:**

* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: talisker/metrics.py [hunk 13]
```diff
@@ -132,3 +137,180 @@ def inc(self, amount=1, **labels):
...
+        if metric.type == 'histogram':
+            sink = histograms
+        elif metric.type == 'counter':
+            sink = counters
+
+        for name, labels, value in metric.samples:
+            key = json.dumps(
+                (metric.name, name, tuple(labels), tuple(labels.values()))
+            )
+            sink.write_value(key, value)
```
[the remark this answers is not in the document]
Comment (bloodearnest) [reply talisker/metrics.py]: hmm, probably, although there various fd's should be closed when they go out of scope.

But will close them explicitly

**Comment by bloodearnest** [comment]
This pull request **introduces 1 alert** and **fixes 1** when merging 5cfa09854d0b6bf8c6e48caf2998c0a5b2b267f4 into dabd40b8b4a5a2a630661cbad960ca18bdfb5143 - [view on LGTM.com](https://lgtm.com/projects/g/canonical-ols/talisker/rev/pr-322a207ae3261226af2383ce417a1b78074ae3df)

**new alerts:**

* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

**fixed alerts:**

* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**